packages/react/src/utils.ts

```diff
@@ -0,0 +1,173 @@
+import { GadgetError, gadgetErrorFor, getNonNullableError, InvalidFieldError } from "@gadgetinc/api-client-core";
+import type { CombinedError, OperationResult } from "@urql/core";
+import { GraphQLError } from "graphql";
+import { Operation, OperationContext, UseQueryState } from "urql";
+
+/**
+ * The inner result object returned from a query result
+ **/
+export interface ReadHookState<Data = any, Variables = Record<string, any>> {
+  fetching: boolean;
+  stale: boolean;
+  data?: Data;
+  error?: ErrorWrapper;
+  extensions?: Record<string, any>;
+  operation?: Operation<Data, Variables>;
+}
+
+/**
+ * The return value of a `useGet`, `useFindMany`, `useFindOne` etc hook.
+ * Includes the data result object and a refetch function.
+ **/
+export declare type ReadHookResult<Data = any, Variables = Record<string, any>> = [
+  ReadHookState<Data, Variables>,
+  (opts?: Partial<OperationContext>) => void
+];
+
+/**
+ * The inner result object returned from a mutation result
+ */
+export interface ActionHookState<Data = any, Variables = Record<string, any>> {
+  fetching: boolean;
+  stale: boolean;
+  data?: Data;
+  error?: ErrorWrapper;
+  extensions?: Record<string, any>;
+  operation?: Operation<Data, Variables>;
+}
+
+/**
+ * The return value of a `useAction`, `useGlobalAction`, `useBulkAction` etc hook.
+ * Includes the data result object and a function for running the mutation.
+ **/
+export declare type ActionHookResult<Data = any, Variables = Record<string, any>> = [
+  ActionHookState<Data, Variables>,
+  (variables?: Variables, context?: Partial<OperationContext>) => Promise<OperationResult<Data, Variables>>
+];
+
+const generateErrorMessage = (networkErr?: Error, graphQlErrs?: GraphQLError[]) => {
+  let error = "";
+  if (networkErr !== undefined) {
+    return (error = `[Network] ${networkErr.message}`);
+  }
+
+  if (graphQlErrs !== undefined) {
+    graphQlErrs.forEach((err) => {
+      error += `[GraphQL] ${err.message}\n`;
+    });
+  }
+
+  return error.trim();
+};
+
+const rehydrateGraphQlError = (error: any): GraphQLError => {
+  if (typeof error === "string") {
+    return new GraphQLError(error);
+  } else if (typeof error === "object" && error.message) {
```

Comment:
`typeof null == "object"`, so I'd guard with `error?.message` (or some other thing to prevent accessing a prop on `null`).